Device Grant according to RFC8628

KirschbaumP opened the pull request:
I created the device grant according to [RFC8628](https://tools.ietf.org/html/rfc8628).

jonasfj commented on lib/src/device_authorization_grant.dart:
```diff
@@ -0,0 +1,385 @@
+  final String _delimiter;
+
+  /// The HTTP client used to make HTTP requests.
+  http.Client? _httpClient;
```
Why is this nullable? the constructor doesn't seem to allow it.

KirschbaumP replied:
fixed by commit [168a984](https://github.com/dart-lang/oauth2/pull/112/commits/168a9847c5ede4e4a335846e65ae70cd782ae302)

jonasfj commented:
Maybe, we should:
 * Make an example that shows how to do this using [sign-in with google](https://developers.google.com/identity/protocols/oauth2/limited-input-device), so that it's easy to test that this actually works.
 * Write some tests...

----------------------

I'm not oppose to landing this. But I need to convinced that it works, an easy to configure example for using this with [sign-in with google](https://developers.google.com/identity/protocols/oauth2/limited-input-device) would convince me, I think :D

--------------

Also update changelog.

KirschbaumP commented:
> Maybe, we should:
> 
> * Make an example that shows how to do this using [sign-in with google](https://developers.google.com/identity/protocols/oauth2/limited-input-device), so that it's easy to test that this actually works.
> * Write some tests...
> 
> I'm not oppose to landing this. But I need to convinced that it works, an easy to configure example for using this with [sign-in with google](https://developers.google.com/identity/protocols/oauth2/limited-input-device) would convince me, I think :D
> 
> Also update changelog.
@jonasfj I've added some test, updated the readme and changelog.

I also tested it with a keycloak server and with google login.
To test it with google signin you have to create a new client (Application type: "TVs and Limited Input Devices") and set the client id and secret. Thats all :)

jonasfj commented on pubspec.yaml:
```diff
@@ -1,5 +1,5 @@
 name: oauth2
-version: 2.0.0
+version: 2.0.1
```
```suggestion
version: 2.1.0
```

jonasfj commented on lib/src/device_authorization_grant.dart:
```diff
@@ -0,0 +1,401 @@
+
+    var startTime = DateTime.now();
+
+    var response =
+        await _httpClient.post(deviceEndpoint, headers: headers, body: body);
```
Does the RFC specify any retry logic? If not is there a retry logic that we should consider implementing?

Won't most users of something like this want to have retry logic around the HTTP request?

KirschbaumP replied:
The RFC talks about polling. A standard interval of 5 seconds (according to RFC) is assumed, unless otherwise defined.

KirschbaumP replied:
I have implemented a basic logic for the query, but the user can also implement his own query logic

jonasfj commented on lib/src/device_authorization_grant.dart:
```diff
@@ -0,0 +1,401 @@
+
+        var description = parameters['error_description'];
+        var uriString = parameters['error_uri'];
+        var uri = uriString == null ? null : Uri.parse(uriString);
```
Consider try/catch around `Uri.parse` and throw a new `FormatException` with a custom message.

KirschbaumP replied:
I would not trigger an error here, because only the error message is parsed here. If the error_uri is not valid I would rather throw an DeviceException with an empty uri than a FormatException.

jonasfj commented on lib/src/device_authorization_grant.dart:
```diff
@@ -0,0 +1,401 @@
+
+  /// starts the Flow and returns [DeviceProperties]
+  /// [useSecret] whether to use the clientSecret for the token-request or not
+  Future<DeviceProperties> getDeviceCode(
```
We throw `Exception` when users are expected to sometimes want to handle them... Certainly we should document, that this throw `FormatException` if the response from the server invalid.

jonasfj commented on lib/src/device_authorization_grant.dart:
```diff
@@ -0,0 +1,401 @@
+
+  /// Check if the user has confirmed the device and then return a [Client]
+  /// [useSecret] whether to use the clientSecret for the token-request or not
+  Future<Client> pollForToken({bool useSecret = true}) async {
```
How does this work if the user haven't signed in yet?

Will it throw an `Exception`? If so we should document what it throws?

jonasfj replied:
Perhaps, it's better to return `null` or throw a very specific exception like `DeviceAuthorizationPendingException` ?

Or maybe we should just do: `grant.pollForToken({bool useSecret = true, Duration timeout = const Duration(Minutes: 5)})`, and then have the function repeatedly polling the server for up to 5 minutes (or whatever default we think is reasonable). That way the user of this library won't have to implement a polling loop themselves.

And a user who wants to implement their own polling loop can just do `grant.pollForToken(timeout: Duration())`.

Polling with `device_code.interval ?? 10` seems very reasonable, and if we implement the polling logic here, odds are that users are more likely implement the logic correctly :D

I'm okay, with not putting the polling logic in here too... WDYT?

KirschbaumP replied:
I think it's a good idea to implement the query logic and allow experienced users to implement their own logic.

jonasfj commented on example/device_code_flow.dart:
```diff
@@ -0,0 +1,88 @@
+  while (true) {
+    try {
+      return await grant.pollForToken();
+    } catch (e) {
```
So if `grant.pollForToken()` throws a `StateError` due to a programmer error, then this becomes an infinite loop?

KirschbaumP replied:
should be fixed in f25abe5
Now I throw back every error except AuthorizationExceptions

jonasfj commented on test/device_authorization_grant_test.dart:
```diff
@@ -0,0 +1,141 @@
+      });
+
+      expect(
+          await grant.pollForToken(),
```
Don't we need a test for what happens if `grant.pollForToken()` doesn't succeed, because the user haven't authenticated the device yet... and the application is supposed to call `grant.pollForToken()` again.

jonasfj commented on lib/src/device_code_exception.dart:
```diff
@@ -0,0 +1,39 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// An exception raised when OAuth2 device fails.
+class DeviceException implements Exception {
```
Documentation link:
```suggestion
/// An exception raised when [DeviceAuthorizationGrant] fails.
class DeviceException implements Exception {
```

Can we be more specific about from where and when this is thrown... In what cases...

KirschbaumP replied:
should be fixed in f25abe5

jonasfj reviewed (commented):
Sorry, about the turn around... I think we need to testing of the polling loop... and possibly consider owning that part.. at-least Exceptions thrown  should be documented.

Albert221 commented on README.md:
````diff
@@ -240,6 +240,90 @@ var result = await client.read('http://example.com/protected-resources.txt');
 File('~/.myapp/credentials.json').writeAsString(client.credentials.toJson());
 ```
 
+## Deivce Code Grant
````
```suggestion
## Device Code Grant
```

mosuem commented:
Closing this, as it is stale and the files are out of date. Feel free to reopen in case you still want to land it!
